services/imagery/src/backends/sync-backend.js

```diff
@@ -1,108 +1,94 @@
-
-                // Throw an error if the id number went down.
-                if (latestId < lastId)
-                    throw Error('Unexpected latest id number');
-
-                // Getting the images we don't have.
-                for (let id = lastId + 1; id <= latestId; id++) {
-                    console.log('Fetching image: ' + id);
-
-                    let msg = await this._getImage(id);
-
-                    // Getting the image data from the message.
-                    let image = msg.image;
-
-                    // Clearing the image data so only metadata is
-                    // left.
-                    msg.image = null;
-
-                    // Adding it to the image store.
-                    await this._imageStore.addImage(image, msg);
-                }
-
-                lastId = latestId;
-            } catch (err) {
-                let message = err.name + ': ' + err.message;
-                console.error('Encountered an error in sync loop: ' + message);
-            }
-
-            // Wait 250 ms to continue.
-            await wait(250);
-        }
-    }
-
-    /** Get the latest id number the sync url has. */
-    async _getLatestId() {
-        let msg = await request({
-            uri: `http://${this._syncUrl}/api/count`,
-            encoding: null,
-            transform: buffer => imagery.ImageCount.decode(buffer),
-            transform2xxOnly: true,
-            timeout: 5000
-        });
+addProtobuf(request);
 
-        return msg.count - 1;
-    }
-
-    /** Get an image from the sync url by id. */
-    async _getImage(id) {
-        return await request({
-            uri: `http://${this._syncUrl}/api/image/${id}`,
-            qs: {
-                // Making sure we have both the image and warped
-                // image if it's available.
-                original: 'true',
-                warped: 'true'
-            },
-            encoding: null,
-            transform: buffer => imagery.Image.decode(buffer),
-            transform2xxOnly: true,
-            timeout: 5000
+export default class SyncBackend {
+  /**
+   * Backend which pulls data from another imagery service.
+   *
+   * This is useful for mirroring imagery data on another machine
+   * from where the imagery originates from.
+   *
+   * Telemetry data is simply gathered from the original image.
+   */
+
+  /** Create a new sync backend. */
+  constructor(imageStore, syncUrl) {
+    this._imageStore = imageStore;
+    this._syncUrl = syncUrl;
+    this._active = false;
+  }
+
+  /** Start the sync loop in the background. */
+  async start() {
+    this._active = true;
+    this._runLoop();
+  }
+
+  async stop() {
+    await this._task.stop();
+  }
+
+  async _runLoop() {
+    // The last image number we've fetched.
+    // Making it into a set allows constant-time lookup.
+    let stored = new Set(await this._imageStore.getAvailable());
+
+    this._task = createTimeoutTask(async () => {
```

Comment:
Can't be a timeout task because it might cause duplicate requests if downloading is still occurring after the interval. Instead, poll periodically when there are no images left to download.

Reply:
Timeout tasks waits until task is done. I'm dum